src/cassandra/future.rs

```diff
@@ -1,333 +1,254 @@
-use Session;
-
 use cassandra::error::{CassError, CassErrorResult};
 use cassandra::prepared::PreparedStatement;
 use cassandra::result::CassResult;
 use cassandra::util::Protected;
+use cassandra_sys::CassError_;
 use cassandra_sys::CASS_OK;
-
 use cassandra_sys::CassFuture as _Future;
-use cassandra_sys::CassFutureCallback as _CassFutureCallback;
-use cassandra_sys::cass_future_custom_payload_item;
-use cassandra_sys::cass_future_custom_payload_item_count;
 use cassandra_sys::cass_future_error_code;
 use cassandra_sys::cass_future_error_message;
-// use cassandra_sys::CassResult as _CassResult;
 use cassandra_sys::cass_future_free;
 use cassandra_sys::cass_future_get_error_result;
 use cassandra_sys::cass_future_get_prepared;
 use cassandra_sys::cass_future_get_result;
 use cassandra_sys::cass_future_ready;
 use cassandra_sys::cass_future_set_callback;
-use cassandra_sys::cass_future_wait;
-use cassandra_sys::cass_future_wait_timed;
-
 use cassandra_sys::cass_true;
+
 use errors::*;
 use std::mem;
-use std::os::raw;
 use std::slice;
 use std::str;
-
-/// A CQL Future representing the status of any asynchronous calls to Cassandra
+use std::sync::{Arc, Mutex};
+use std::marker::PhantomData;
+use futures;
+
+/// A future representing the result of a Cassandra driver operation.
+///
+/// On success, returns a result of type `T`. On failure, returns a Cassandra error.
+///
+/// Take care only to construct this at the appropriate type (i.e., where the
```

Comment:
Hmm, this is my programming language theory background showing, I fear - I probably need to reword but not sure how.

In `CassFuture<CassResult>`, it's not just that the `CassFuture` is constructed using the type `CassResult` (i.e., you supply that type at construction time), but that the future itself is now qualified by that type, so everything I do with the `CassFuture` now yields the `CassResult`-specific behaviour - I have a `CassFuture` at type `CassResult`.

I'll have a think.